### PR TITLE
improve pinning projection docs

### LINKED ISSUE #61272: Improve pin projection documentation
In [this forum thread](https://users.rust-lang.org/t/when-is-it-safe-to-move-a-member-value-out-of-a-pinned-future/28182), @shepmaster 	wonders about the details of pinning projections, in particular when there are multiple fields that can make *different* choices for being structural or not.  The docs could likely be improved for that.

This is just so that we don't forget.

### PR DESCRIPTION
**Pull request by RalfJung**
This tries to improve the explanation of structural pinning and pinning projections based on [this URLO thread](https://users.rust-lang.org/t/when-is-it-safe-to-move-a-member-value-out-of-a-pinned-future/28182).

Fixes https://github.com/rust-lang/rust/issues/61272.

### CODE REVIEW

**Comment by RalfJung** [comment]
Cc @Centril you usually have a bunch of feedback on these ;)

File: src/libcore/pin.rs [hunk 1]
```diff
@@ -166,57 +167,128 @@
...
+//!
+//! It turns out that it is actually up to the author of the data structure
+//! to decide whether the pinned projection for a particular field turns
+//! `Pin<&mut Struct>` into `Pin<&mut Field>` or `&mut Field`. There are some
+//! constraints though, and the most important constraint is *consistency*:
+//! every field can be *either* projected to a pinned reference, *or* have
+//! pinning removed as part of the projection. If both are done for the same field,
+//! that will likely be unsound!
+//!
+//! Basically, as the author of a data structure you get to decide for each field whether pinning
```
Comment (Centril) [inline src/libcore/pin.rs]:
```suggestion
//! As the author of a data structure you get to decide for each field whether pinning
```

File: src/libcore/pin.rs [hunk 2]
```diff
@@ -166,57 +167,128 @@
...
+//! "propagates" to this field or not. Pinning that propagates is also called "structural",
+//! because it follows the structure of the type.
+//! In the following, we describe the considerations that have to be made for either choice.
```
Comment (Centril) [inline src/libcore/pin.rs]:
```suggestion
//! In the following subsection, we describe the considerations that have to be made for either choice.
```

File: src/libcore/pin.rs [hunk 3]
```diff
@@ -166,57 +167,128 @@
...
+//!
+//! ## Pinning *is not* structural for `field`
+//!
+//! It may seem counter-intuitive that the field of a pinned struct might not be pinned,
+//! but that is actually the easiest choice: if a `Pin<&mut Field>` is never created,
+//! nothing can go wrong! So, if you decide that some field does not have structural pinning,
+//! all you have to ensure is that you never create a pinned reference to that field.
+//!
+//! Then you may add a projection method that turns `Pin<&mut Struct>` into `&mut Field`:
```
Comment (Centril) [inline src/libcore/pin.rs]: This sentence should be joined with the previous paragraph.
I'd also start with "If you've done that, you may..."

File: src/libcore/pin.rs [hunk 4]
```diff
@@ -166,57 +167,128 @@
...
+//!         unsafe { &mut self.get_unchecked_mut().field }
+//!     }
+//! }
+//! ```
 //!
-//! Having a pinning projection for some field means that pinning is "structural":
-//! when the wrapper is pinned, the field must be considered pinned, too.
-//! After all, the pinning projection lets us get a `Pin<&Field>`.
+//! You may also `impl Unpin for Struct` *even if* the type of `field`
+//! is not `Unpin`.  What that type thinks about pinning is just not relevant
```
Comment (Centril) [inline src/libcore/pin.rs]:
```suggestion
//! is not `Unpin`. What that type thinks about pinning is not relevant
```

**Comment by Centril** [comment]
@RalfJung Looks good aside from a few nits ^---

File: src/libcore/pin.rs (hunk 2 above)
Comment (RalfJung) [reply src/libcore/pin.rs]: Shouldn't this be "subsections"?

File: src/libcore/pin.rs (hunk 3 above)
Comment (RalfJung) [reply src/libcore/pin.rs]: There are two separate things I am talking about here -- the high-level idea, and then the first consequence (adding the projection). I think they should be in different paragraphs.

File: src/libcore/pin.rs (hunk 2 above)
Comment (Centril) [reply src/libcore/pin.rs]: Yeah; I left this note when I hadn't read all the sections :)

File: src/libcore/pin.rs (hunk 3 above)
Comment (Centril) [reply src/libcore/pin.rs]: It seems odd that the antecedent "have ensured that..." is not in the same paragraph as the consequent. I would suggest not beginning the sentence with "Then..." if you want them in different paragraphs as starting that way misses the "If...".

Comment (RalfJung) [reply src/libcore/pin.rs]: What about this?
```
//! Fields without structural pinning may have a projection method that turns
//! `Pin<&mut Struct>` into `&mut Field`:
```

Comment (Centril) [reply src/libcore/pin.rs]: Looks good :)

**Comment by RalfJung** [comment]
@rust-lang/libs this still awaits review.

**Comment by RalfJung** [comment]
@withoutboats could you maybe help review this?

**Review by Dylan-DPC-zz — APPROVED** [review]

**Pull request merged**